onnxruntime/__init__.py

```diff
@@ -76,3 +76,97 @@
     __version__ = version
 
 onnxruntime_validation.check_distro_info()
+
+
+def check_and_load_cuda_libs(root_directory, cuda_libs_):
+    # Convert the target library names to lowercase for case-insensitive comparison
+    # Convert the target library names to lowercase for case-insensitive comparison
+    if cuda_libs_ is None or len(cuda_libs_) == 0:
+        logging.info("No CUDA libraries provided for loading.")
+        return
+    cuda_libs_ = {lib.lower() for lib in cuda_libs_}
+    found_libs = {}
+    for dirpath, _, filenames in os.walk(root_directory):
+        # Convert filenames in the current directory to lowercase for comparison
+        files_in_dir = {file.lower(): file for file in filenames}  # Map lowercase to original
+        # Find common libraries in the current directory
+        matched_libs = cuda_libs_.intersection(files_in_dir.keys())
+        for lib in matched_libs:
+            # Store the full path of the found DLL
+            full_path = os.path.join(dirpath, files_in_dir[lib])
+            found_libs[lib] = full_path
+            try:
+                # Load the DLL using ctypes
+                _ = ctypes.CDLL(full_path)
+                logging.info(f"Successfully loaded: {full_path}")
+            except OSError as e:
+                logging.info(f"Failed to load {full_path}: {e}")
+
+        # If all required libraries are found, stop the search
+        if set(found_libs.keys()) == cuda_libs_:
+            logging.info("All required CUDA libraries found and loaded.")
+            return
+    logging.info(
+        f"Failed to load CUDA libraries from site-packages/nvidia directory: {cuda_libs_ - found_libs.keys()}. They might be loaded later from standard search paths for shared libraries."
+    )
+    return
+
+
+# Load nvidia libraries from site-packages/nvidia if the package is onnxruntime-gpu
+if cuda_version is not None and cuda_version != "":
+    import ctypes
+    import logging
+    import os
+    import platform
+    import site
+
+    cuda_version_ = tuple(map(int, cuda_version.split(".")))
+    # Get the site-packages path where nvidia packages are installed
+    site_packages_path = site.getsitepackages()[-1]
```

Comment:
Shouldn't we check all sitepackages directories? We might also check whether things like `nvidia.cudnn` and `nvidia.cudnn.__path__` are importable (`import nvidia.cudnn` or importlib)